v2/uuid/pgxcompat/uuid.go

```diff
@@ -0,0 +1,258 @@
+	case []byte:
+		if value == nil {
+			*u = UUID{Status: pgtype.Null}
+		} else if uuidStr, err := fromBinary(value[:]); err != nil {
+			return err
+		} else {
+			*u = UUID{UUID: uuid.UUID(uuidStr), Status: pgtype.Present}
+		}
+	case string:
+		if len(value) == uuidStrShortLength {
+			value = fmt.Sprintf("%s-%s-%s-%s-%s", value[0:8], value[8:12], value[12:16], value[16:20], value[20:32])
+		}
+
+		*u = UUID{UUID: uuid.UUID(value), Status: pgtype.Present}
+	case *string:
+		if value == nil {
+			*u = UUID{Status: pgtype.Null}
+		} else {
+			return u.Set(*value)
+		}
+	default:
+		return fmt.Errorf("cannot convert %v of type %T to UUID", value, value)
+	}
+
+	return nil
+}
+
+func (u UUID) Get() interface{} {
+	switch u.Status {
+	case pgtype.Present:
+		return u.UUID
+	case pgtype.Null:
+		return nil
+	default:
+		return u.Status
+	}
+}
+
+func (u *UUID) AssignTo(target interface{}) (err error) {
+	switch u.Status {
+	case pgtype.Present:
+		switch dst := target.(type) {
+		case *uuid.UUID:
+			*dst = u.UUID
+			return nil
+		case *[16]byte:
+			*dst, err = toBinary(u.UUID.String())
+			return err
+		case *[]byte:
+			*dst = make([]byte, uuidBinaryLength)
+			bin, err := toBinary(u.UUID.String())
+			copy(*dst, bin[:])
+
+			return err
+		case *string:
+			*dst = u.UUID.String()
+			return nil
+		default:
+			if nextDst, retry := pgtype.GetAssignToDstType(dst); retry {
+				return u.AssignTo(nextDst)
+			}
+
+			return fmt.Errorf("unable to assign to %T", dst)
+		}
+	case pgtype.Null:
+		return pgtype.NullAssignTo(target)
+	}
+
+	return fmt.Errorf("cannot assign %v into %T", u, target)
+}
+
+func (u *UUID) DecodeText(ci *pgtype.ConnInfo, value []byte) error {
+	if value == nil {
+		*u = UUID{Status: pgtype.Null}
+		return nil
+	}
+
+	uuidValue := uuid.UUID(string(value))
+	if err := uuidValue.Validate(); err != nil {
+		return err
+	}
+
+	*u = UUID{UUID: uuidValue, Status: pgtype.Present}
+
+	return nil
+}
+
+func (u *UUID) DecodeBinary(ci *pgtype.ConnInfo, value []byte) error {
+	if value == nil {
+		*u = UUID{Status: pgtype.Null}
+		return nil
+	}
+
+	uuidValue, err := fromBinary(value)
+	if err != nil {
+		return err
+	}
+
+	*u = UUID{UUID: uuid.UUID(uuidValue), Status: pgtype.Present}
+
+	return nil
+}
+
+func (u UUID) EncodeText(ci *pgtype.ConnInfo, outBuf []byte) ([]byte, error) {
+	switch u.Status {
+	case pgtype.Null:
+		return nil, nil
+	case pgtype.Undefined:
+		return nil, errUndefined
+	}
+
+	return append(outBuf, u.UUID.String()...), nil
+}
+
+func (u UUID) EncodeBinary(ci *pgtype.ConnInfo, outBuf []byte) ([]byte, error) {
+	switch u.Status {
+	case pgtype.Null:
+		return nil, nil
+	case pgtype.Undefined:
+		return nil, errUndefined
+	}
+
+	tmpBuf, err := toBinary(u.UUID.String())
+	if err != nil {
+		return nil, err
+	}
+
+	return append(outBuf, tmpBuf[:]...), nil
+}
+
+// Scan implements the database/sql Scanner interface.
+func (u *UUID) Scan(value interface{}) error {
```

Comment:
Could we test that pgxcompat.UUID correctly implements the ínterfaces that it is supposed to? For instance the scanner interface here

Reply:
It is not obvious to me which of these methods are required by pgx interfaces and/or which are created by ourselves for utility.

Reply:
Found one [here](https://github.com/jackc/pgtype/blob/90af821478c74fd8f917b7301d79886f933e5fe3/pgtype.go#L127)

Reply:
Added test so that we implement [`pgtype.ValueTranscoder`](https://github.com/jackc/pgtype/blob/90af821478c74fd8f917b7301d79886f933e5fe3/pgtype.go#L157l)